prometheus/client.go

```diff
@@ -0,0 +1,352 @@
+	if err != nil {
+		return err
+	}
+
+	// only lock when we are sure we are going to change the instance of the sub client
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	c.url = prometheusURL
+	if isCompatible {
+		c.subClient = &compatibleHTTPClient{
+			requestTimeout:   c.requestTimeout,
+			prometheusClient: v1.NewAPI(prometheusHTTPClient),
+		}
+	} else {
+		c.subClient = &notCompatibleHTTPClient{
+			requestTimeout:   c.requestTimeout,
+			prometheusClient: v1.NewAPI(prometheusHTTPClient),
+		}
+	}
+
+	return nil
+}
+
+func (c *httpClient) isCompatible(prometheusURL string) (bool, error) {
+	httpRequest, err := buildStatusRequest(prometheusURL)
+	if err != nil {
+		return false, err
+	}
+	httpClient := &http.Client{
+		Transport: buildGenericRoundTripper(c.requestTimeout * time.Second),
+		Timeout:   c.requestTimeout * time.Second,
+	}
+	resp, err := httpClient.Do(httpRequest)
+	if err != nil {
+		return false, err
+	}
+
+	// For prometheus version less than 2.14 `api/v1/status/buildinfo` was not supported this can
+	// break many function which solely depends on version comparing like `hover`, etc.
+	if resp.StatusCode == http.StatusNotFound {
+		return false, nil
+	}
+	defer resp.Body.Close() // nolint: errcheck
+	if resp.Body != nil {
+		data, err := ioutil.ReadAll(resp.Body)
+		jsonResponse := buildInfoResponse{}
+		err = json.Unmarshal(data, &jsonResponse)
+		if err != nil {
+			return false, err
+		}
+		currentVersion, err := semver.New(jsonResponse.Data.Version)
+		if err != nil {
+			return false, err
+		}
+		return currentVersion.GTE(requiredVersion), nil
+	}
+	return false, nil
+}
+
+// compatibleHTTPClient must be used to contact a distant prometheus with a version >= v2.15
+type compatibleHTTPClient struct {
+	Client
+	requestTimeout   time.Duration
+	prometheusClient v1.API
+}
+
+func (c *compatibleHTTPClient) Metadata(metric string) (v1.Metadata, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), c.requestTimeout*time.Second)
+	defer cancel()
+
+	metadata, err := c.prometheusClient.Metadata(ctx, metric, "1")
+	if err != nil {
+		return v1.Metadata{}, err
+	}
+	if len(metadata) <= 0 {
+		return v1.Metadata{}, nil
+	}
+	return v1.Metadata{
+		Type: metadata[metric][0].Type,
+		Help: metadata[metric][0].Help,
+		Unit: metadata[metric][0].Unit,
+	}, nil
+}
+
+func (c *compatibleHTTPClient) AllMetadata() (map[string][]v1.Metadata, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), c.requestTimeout*time.Second)
+	defer cancel()
+	return c.prometheusClient.Metadata(ctx, "", "")
+}
+
+func (c *compatibleHTTPClient) LabelNames(name string) ([]string, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), c.requestTimeout*time.Second)
+	defer cancel()
+	if len(name) <= 0 {
+		names, _, err := c.prometheusClient.LabelNames(ctx)
+		return names, err
+	}
+	labelNames, _, err := c.prometheusClient.Series(ctx, []string{name}, time.Now().Add(-100*time.Hour), time.Now())
```

Comment:
It might be a good idea to make the look back (100h) configurable.

Reply:
for this last part, yeah why not. But do you want it configurable from the configuration file ?

Reply:
Ideally it would be configurable both from the configuration file and from https://github.com/prometheus-community/promql-langserver/blob/8a91c484c900328c3953d5fb9832b4f0a65f7dc9/langserver/config.go#L56 . But that can be deferred to an independent PR.

Reply:
let's see in a different PR, I'm not yet sure how to handle it properly